File: leetcode/july-challenge/week-3/top_k_freq_elements.cxx
struct cmp {
    bool operator()(const pair<int, int>& p1, const pair<int, int>& p2) {
        return p1.first < p2.first;
    }
};

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        const int n = nums.size();
        unordered_map<int, pair<int, int>> umap;
        priority_queue<pair<int, int>, vector<pair<int, int>>, cmp> pq;
        vector<int> res;
        
        for (int i = 0; i < n; ++i) {
            pair<int, int> p;
            if (umap.find(nums[i]) == umap.end()) {
                p.first = 1;
                p.second = nums[i];
            } else {
                p = umap[nums[i]];
                ++p.first;
            }
            umap[nums[i]] = p;
        }
        
        for (auto it = umap.begin(); it != umap.end(); ++it) {
            pq.push(it->second);
        }
        
        for (int i = 0; i < k; ++i) {
            pair<int, int> p = pq.top();
            pq.pop();
            res.push_back(p.second);
        }
        return res;
    }
};

File: leetcode/july-challenge/week-3/TopKFreqElements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        Queue<Integer> pq;
        int[] res = new int[k];
        
        if (k == nums.length) {
            return nums;
        }
        
        for (int i : nums) {
            map.put(i, map.getOrDefault(i, 0) + 1);
        }
        
        pq = new PriorityQueue<>((n1, n2) -> map.get(n1) - map.get(n2));
        
        for (int n : map.keySet()) {
            pq.add(n);
            if (pq.size() > k) {
                pq.poll();
            }
        }
        
        for (int i = k - 1; i >= 0; --i) {
            res[i] = pq.poll();
        }
        
        return res;
    }
}
